File: 백준/Gold/2252. 줄 세우기/줄 세우기.java
import java.io.*;
import java.util.*;

public class Main {

    static int n;
    static int m;
    static List<ArrayList<Integer>> adj = new ArrayList<>();
    static int[] indegree = new int[32002];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        for (int i = 0; i <= n; i++) {
            adj.add(new ArrayList<>());
        }
        while (m-- > 0) {
            st = new StringTokenizer(br.readLine());
            int u = Integer.parseInt(st.nextToken());
            int v = Integer.parseInt(st.nextToken());
            adj.get(u).add(v);
            indegree[v]++;
        }
        Queue<Integer> queue = new ArrayDeque<>();
        for (int i = 1; i <= n; i++) {
            if (indegree[i] == 0) queue.offer(i);
        }

        StringBuilder sb = new StringBuilder();
        while (!queue.isEmpty()) {
            int cur = queue.poll();
            sb.append(cur).append(' ');

            for (int nxt : adj.get(cur)) {
                indegree[nxt]--;
                if (indegree[nxt] == 0) queue.offer(nxt);
            }
        }

        System.out.print(sb);
    }
}

File: 백준/Gold/2252. 줄 세우기/줄 세우기.kt
import java.util.*

fun main() = with(System.`in`.bufferedReader()) {
    val (n, m) = readLine().split(" ").map { it.toInt() }
    val adj = Array(n + 1) { mutableListOf<Int>() }
    val indeg = IntArray(n + 1)

    repeat(m) {
        val st = StringTokenizer(readLine())
        val u = st.nextToken().toInt()
        val v = st.nextToken().toInt()
        adj[u] += v
        indeg[v]++
    }

    val queue = ArrayDeque<Int>()
    for (i in 1..n) if (indeg[i] == 0) queue.offer(i)

    val sb = StringBuilder()
    while (queue.isNotEmpty()) {
        var cur = queue.poll()
        sb.append(cur).append(' ')

        for (nxt in adj[cur]) {
            if (--indeg[nxt] == 0) queue.offer(nxt)
        }
    }

    print(sb)
}